### ISSUE TITLE
Use Application.executeOnPooledThread rather than Executors

### ISSUE THREAD

**Issue by baron1405** [jshiell/checkstyle-idea#239]
In CheckerFactoryCache the cleanUpExecutor is using the Java Executors class to schedule threads from a pool. Unfortunately, this leads to leaked threads. Under normal execution this is not really a problem. However, in unit tests derived from com.intellij.testFramework.PlatformTestCase, the leaked threads cause the tests to fail due to the ThreadTracker.checkLeak() detection being called during tearDown.

Instead of using the Java Executors class, use the Application.executeOnPooledThread method to schedule your cleanup thread (and any other threads). This allows IntelliJ to manage the thread pool and prevents threads from being leaked.

To reproduce this issue, create a unit test derived from PlatformTestCase and just have the CheckStyle-IDEA plugin as a dependency of the test (i.e. installed in the plugin sandbox). The test can just do an assertTrue(true). Just the loading of the CheckStyle-IDEA plugin during test runtime will cause the cleanup thread to be scheduled and leaked. The test should fail with a stack trace similar to:

```
java.lang.AssertionError: Thread leaked: Thread[pool-2-thread-1,5,main]; TIMED_WAITING (true)
--- its stacktrace:
 at sun.misc.Unsafe.park(Native Method)
 at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
 at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
 at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1093)
 at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:809)
 at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1067)
 at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1127)
 at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
 at java.lang.Thread.run(Thread.java:745)
---

	at org.junit.Assert.fail(Assert.java:88)
	at com.intellij.testFramework.ThreadTracker.checkLeak(ThreadTracker.java:161)
	at com.intellij.testFramework.PlatformTestCase.tearDown(PlatformTestCase.java:494)
	at com.intellij.testFramework.PlatformTestCase$5.run(PlatformTestCase.java:679)
	at com.intellij.testFramework.EdtTestUtil$Companion$runInEdtAndWait$1.invoke(EdtTestUtil.kt:26)
	at com.intellij.testFramework.EdtTestUtil$Companion$runInEdtAndWait$1.invoke(EdtTestUtil.kt:24)
	at com.intellij.testFramework.EdtTestUtilKt$sam$Runnable$ca4818df.run(EdtTestUtil.kt)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:301)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:749)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:702)
	at java.awt.EventQueue$3.run(EventQueue.java:696)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:75)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:719)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:366)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```

**Comment by jshiell** [comment]
Thanks for the report. I'll have a look, but it won't be immediately due to my country committing economic and political suicide.

**Comment by baron1405** [comment]
I completely understand :-( Unfortunately, we will have the opportunity to make a similar error of judgement in November. Hopefully, sanity will prevail eventually for all of us.

**Comment by jshiell** [comment]
Fingers crossed! In any case, I'll try have a look later this week once I'm done moping 😄

**Comment by baron1405** [comment]
I just created pull request #242 as a possible way to have IntelliJ manage the cleanup thread. Note that the executeOnPooledThread suggestion above is not really what we need. Instead I looked for an IntelliJ platform way to get a schedule thread executor. Found it in the JobScheduler. This has some trade offs that are described in the pull request. Hopefully, this gives you a head start on a solution.
